### ISSUE TITLE
throw indicator

### ISSUE THREAD

**Issue by Mouvedia** [ericelliott/rtype#12]
Something should tell us if a function can throw.

Swift got this built in: https://developer.apple.com/library/prerelease/ios/documentation/Swift/Conceptual/Swift_Programming_Language/ErrorHandling.html#//apple_ref/doc/uid/TP40014097-CH42-ID510

**Comment by ericelliott** [comment]
I agree. Ideas? PR?

**Comment by ericelliott** [comment]
The swift examples are:

``` swift
cannotThrowErrors() -> String
canThrowErrors() throws -> String
```

To me, that looks like it throws a String, which is silly. What if we did this, instead?

``` js
cannotThrowErrors() => String
canThrowErrors throws: TypeError () => String
```

You could even throw multiple types of errors:

``` js
myMiddleware throws: TypeError, throws: NotFoundError () => String
```

and multi-line:

``` js
myMiddleware
  throws: TypeError,
  throws: NotFoundError
  () => String
```

On the fat arrow and uppercase, see #13 & #16

**Comment by Mouvedia** [comment]
@ericelliott I like the fact that you are adding the type of the error.

For the various case Id use brackets.
e.g.  `throws(TypeError, DOMException)` or `throws[TypeError, DOMException]`

Concerning the placement of the throws keyword without a type, which is the main use case imho, I don't have a strong opinion.
Here are some alternatives:

```
canThrowErrors() =throws> String
canThrowErrors() => String | throws
```

The second one reuse the union type separator. I find that one clearer.

**Comment by ericelliott** [comment]
Union type doesn't make sense there to me -- throws is not a type. I also don't think we need new syntax.

For multiple error types, we could use a union type:

``` js
myMiddleware
  throws: TypeError | NotFoundError
  () => String
```

**Comment by Mouvedia** [comment]
@ericelliott +1 reusing syntax idioms is the way to go.

**Comment by ericelliott** [comment]
Cool. Sounds like we're on the same page. Want to do the PR?

**Comment by Mouvedia** [comment]
@ericelliott  Maybe we could have a shorthand for the generic case, e.g.

```
!canThrowErrors() => String
```

#17

**Comment by ericelliott** [comment]
The `!` syntax is a little confusing. I read it as a logical NOT. Maybe if we moved it to the other side:

``` js
canThrowErrors! () => String
```

Equivalent to:

``` js
canThrowErrors throws () => String
```

**Comment by Mouvedia** [comment]
@ericelliott Yes, it should be on the same side. Done.

**Comment by koresar** [comment]
`canThrowErrors () => String, throws`
`canThrowErrors () => String, throws: TypeError | NotFoundError`
or
`canThrowErrors () => String; throws`
`canThrowErrors () => String; throws: TypeError | NotFoundError`

**Comment by ericelliott** [comment]
Thanks @koresar -- is there a reason you prefer this over the alternative we're using?

``` js
myMiddleware
  throws: TypeError | NotFoundError
  () => String
```

I kindof like the `!` shorthand, too. [comment link](https://github.com/ericelliott/rtype/issues/12#issuecomment-158260546)

**Comment by koresar** [comment]
It's ain't readable for people without syntax knowledge.
Remember what's written on the README?

> An intuitive way to describe interfaces for the purposes of documentation, particularly function signatures.

**intuitive**. The exclamation mark is not intuitive in any way. However, it's your project. I'm sure you know what you are doing.

**Comment by Mouvedia** [comment]
I think the exclamation point is intuitive as it relates to something surprising/unannounced/deranging, like a thrown error.

Since I think it should be ordered consistently, the position becomes evident. Furthermore if you are using the keyword, it reads naturally (like a sentence).

**Comment by tomek-he-him** [comment]
Hi, I’m late to the party, so it seems. I totally love how you can be explicit about what your function throws:

``` js
myMiddleware
  throws: TypeError | NotFoundError
  () => String
```

But I don’t like how you don’t _have to_ be explicit:

``` js
functionName! () => String

// which is equivalent to
functionName throws () => String
```

I only really use `throw` when I want the program to crash – as an equivalent of `panic` in _go_ and `Debug.crash` in _Elm_. It’s very useful sometimes, but I don’t use that very often – especially not in web/universal code.

`throw` + `try` + `catch` is a bad practice we shouldn’t encourage. If we introduce short syntax it kind of tells people “Hey! I’m really convenient to use – feel free to use me often!”

But yeah – as it’s already merged, feel free to ignore my opinion.

**Comment by tomek-he-him** [comment]
> ``` js
> myMiddleware
>   throws: TypeError | NotFoundError
>   () => String
> ```

(By the way, @koresar – I don’t think you’d disagree that this syntax is super readable!)

**Comment by tomek-he-him** [comment]
> placement of the throws keyword without a type, which is the main use case imho

@Mouvedia the whole point of this standard is to define types the user will have to handle! If you need to `catch` and interpret a `TypeError` from some API, you’d treat it completely differently than a `NotFoundError { code: Number, whatWentWrong: String }`!

If I were you, I’d completely disallow saying `throws` without a type. I think it misses the point of this spec!

**Comment by Mouvedia** [comment]
> If we introduce short syntax it kind of tells people “Hey! I’m really convenient to use – feel free to use me often!”

The problem is, in some projects all functions act like that. That's the kind of use cases that covers, I wouldn't want to have `throws` everywhere if it's the default behaviour.

**Comment by koresar** [comment]
@tomekwi I do very agree that the syntax you are talking about is very readable. Although, it's readable when multi lined. Single line looks less readable.

```
stringifyError throws: TypeError (error: Error, options: Object) => String
```

Thus I'd prefer.

```
stringifyError (error: Error, options: Object) => String, throws: TypeError
```

Moreover, code is read many more times than written. I would not split function name from it's incoming and returning types too far apart. These two the most important information for code readers. I doubt first thing you read about a function is if it throws or not.

```
stringifyError throws: TypeError (error: Error, options: Object) => String
^^^^^^^^^^^^^^                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
func. name                        incoming                       returning
```

The following is more readable to me.

```
stringifyError (error: Error, options: Object) => String, throws: TypeError
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
func. name      incoming                       returning
```

**Comment by koresar** [comment]
Additionally, here is how the same example looks without the function name:

```
throws: TypeError () => String
```

Don't you find this too confusing? The `TypeError` looks very much like function name. Right?

**Comment by Mouvedia** [comment]
@koresar has a point about the type error being potentially confused with the function name. That's why I originally wanted to use brackets.

Ill list the alternative proposals tonight.

**Comment by tomek-he-him** [comment]
@koresar @Mouvedia  :+1: :+1:

**Comment by ericelliott** [comment]
Since these signatures are already being used for documentation, and documentation is one of the stated goals, I agree that it's important to be explicit. I think we need to get rid of the `!` shortcut.

``` js
// sad to see you go... );
functionName! () => String

// but @tomekwi and @koresar make great points...
functionName() => String, throws TypeError
```

That said, I think it can go without saying:

``` js
functionName() => String, throws // Is the same as:
functionName() => String, throws Error
```

And throwing a generic `Error` type is a totally legit use-case.

**Comment by ericelliott** [comment]
@Mouvedia, @tomekwi 

I've added you both as collaborators. Please continue to use the pull request process for changes.

From now on, for any substantial changes, I'll use PR's, too. =)

Thanks for your great feedback, @koresar -- I hate to admit it, but I was wrong in this case. It is confusing to have throw-before if you omit the function name.

**Comment by tomek-he-him** [comment]
:+1: Thanks!

**Comment by ericelliott** [comment]
Just one problem... the comma separator already has a special meaning.

What if you mixed throw with a polymorphic function?

``` js
interface Collection {
  (items: Array[]) => Array[],
  (items: Object[]) => Object[]
 }
```

This will require look ahead in the parser, but if it finds anything other than `throws` following the whitespace, it can terminate the signature and move on.

I think to a human being, it will be obvious:

``` js
interface Collection {
  (items: Array[]) => Array[], throws: TypeError,
  (items: Object[]) => Object[], throws: TypeError
 }
```

Still looks very readable to me. Any objections?

**Comment by tomek-he-him** [comment]
How about using `|` for polymorphic functions? It’s kinda like a union type anyway.

**Comment by ericelliott** [comment]
Then you'd have to disambiguate:

``` js
interface Collection {
  (items: Array[]) => Array[], throws: TypeError |
  (items: Object[]) => Object[], throws: TypeError
 }
```

Could this mean it throws a `TypeError` or a function with type `(items: Object[]) => Object[], throws: TypeError`?

**Comment by tomek-he-him** [comment]
Good point!

**Comment by koresar** [comment]
There's no need in comma before the "throws". Just omit it.

**Comment by ericelliott** [comment]
Without the comma we don't need a look ahead, but I care more about how it reads to humans than how it reads to the parser. Which one reads better from the perspective of a human being?

**Comment by ericelliott** [comment]
[Let's vote!](https://twitter.com/_ericelliott/status/667833352460767232)

**Comment by ericelliott** [comment]
[Vote!](https://twitter.com/_ericelliott/status/667833352460767232)

Comma before `throws`?

``` js
interface Collection {
  (items: Array[]) => Array[], throws: TypeError,
  (items: Object[]) => Object[], throws: TypeError
 }
```

Or no comma before `throws`?

``` js
interface Collection {
  (items: Array[]) => Array[] throws: TypeError,
  (items: Object[]) => Object[] throws: TypeError
 }
```

**Comment by Mouvedia** [comment]
Here's my take on this:
1. the `!` should be used as a shorthand for `throws`. It's akin to the usage of `?` for optional and no function name can end with `!` (cf [JS variable name validator](https://mothereff.in/js-variables)) so it's completely safe
2. A function named throws is valid hence putting throws just before `()` is not a good idea IMO
3. `throws` without a type should be generic to cover any possible case (not map to something specific to let ppl do their thing)
4. `throws` should be placed somewhere after the function name (I don't have a strong opinion on where exactly)

**Comment by ericelliott** [comment]
Thanks @Mouvedia. I'll address each point:
1. One of the primary reasons for Rtype is to make functions easier to document, and rtypes will be used to automatically generate interface documentation. A reader unfamiliar with rtype can probably figure out most of our syntax. They might not immediately recognize that `!` denotes that a function may throw. It will denote there's something special about it, but not what that special thing is.
2. Agreed.
3. `throws` without a type already covers any possible `Error` being thrown, including all subtypes of `Error` such as `TypeError`, `NotFoundError`, and so on. It does not cover something that is NOT an error, because throwing anything other than an error or error subtype is DEFINITELY a mistake, and we should alert developers about that mistake.
4. Done. =)

**Comment by ericelliott** [comment]
I think we've had enough votes here, and the changes are merged. Thanks everybody! =)

**Comment by Mouvedia** [comment]
> and we should alert developers about that mistake

I see what you are doing there. Enforcing documentation of ill usages of `throw`. It's a wise and consensually accepted practice.

**Comment by ericelliott** [comment]
Yeah. Similar to using a lint tool to enforce semicolons. Technically, you can leave them off, but good adherence to the idiomatic JavaScript style requires them.

If you're opting into a type tool such as rtype, you're explicitly saying, "I care about using correct types for things", so we should warn when the user doesn't use a correct type for things.

I can think of zero valid use-cases where I'd want to throw an error without an error object and stack trace.

Can you think of any valid reasons to _crash your program_ without providing a stack trace for developers to investigate?

**Comment by Mouvedia** [comment]
> Can you think of any valid reasons to crash your program without providing a stack trace for developers to investigate?

Documenting legacy code?

> I've added you both as collaborators.

Thanks.

**Comment by ericelliott** [comment]
> Documenting legacy code?

Ah, well, you can opt in and out of warnings, but if I had a legacy codebase to maintain, I'd want to be alerted about potentially serious problems such as a case where the program may crash without providing a stack trace. Seems like an argument _for_ enforcing `Error`, to me.

Always be refactoring.

**Comment by tomek-he-him** [comment]
> Can you think of any valid reasons to crash your program without providing a stack trace for developers to investigate?

I actually use [git.io/tiny-error](https://git.io/tiny-error) to display a descriptive crash message without a stack trace – directly to the user of a CLI tool or library.

Node or the browser handle displaying the message and exiting the program. I don’t have to care – whether I’m writing a library or an application – whether I’m in sync code or deeply buried in a heap of streams, events and observables. It always just exits and prints.

Stack traces are helpful if something unexpected goes wrong. But as a user of a CLI tool or library I don’t care about internal details. If the error is something predicted by the developer, I prefer to be shown an elegant, readable description of what went wrong, why and what to do next – than a long and cryptic stack trace.

**Comment by koresar** [comment]
Many are familiar with Java syntax of function interface annotation:

``` Java
Integer myMethod(String s) throws IOException
```

See? No comma. :) Developers are absolutely fine with it.

Although, I must admit this topic is starting to dive into a bikeshedding. :)

**Comment by ericelliott** [comment]
64% of 25 respondents prefer the comma. No great reasons not to do it. The vote passes, for now. =)

**Comment by Mouvedia** [comment]
@koresar that's not what they were voting on.

The main case was the throws keyword placed after another keyword (e.g. requires) or `=> Type`.
In the case of `=> Void` which can be omitted, there are no examples in the specification covering that case.

To be consistent with #113, Id support not having a comma before throws **if** the returned type has been omitted and it's the first keyword.

tl;dr: @ericelliott `=>`, `throws`, `requires` and `emits` should be in the same token category grammar-wise.
